core/ClassInfo.php

```diff
@@ -52,17 +52,20 @@ public static function hasTable($class) {
 			return false;
 		}
 	}
-	
+
 	public static function reset_db_cache() {
 		self::$_cache_all_tables = null;
 		self::$_cache_ancestry = array();
 	}
-	
+
 	/**
 	 * Returns the manifest of all classes which are present in the database.
 	 * @param string $class Class name to check enum values for ClassName field
 	 */
 	public static function getValidSubClasses($class = 'SiteTree', $includeUnbacked = false) {
+		if (self::hasTable($class)) {
+			$class = self::$_cache_all_tables[strtolower($class)];
```

Comment:
`hasTable` only calls the DB once and it's a `LIST TABLES` query. I assume this is called at least once every time there's a request and so this isn't adding any extra DB overhead.

I agree accessing that cache is pretty nasty.

An overall comment I'd make is to remember I've tried to do this entire PR in the least intrusive way possible in an attempt to get it into a patch release. In hind-sight I think that there are too many flaws in the Manifest and in `ClassInfo` to make this realistic - so this is probably more a proof of concept now and should be moved to master to allow a greater re-write of the other parts of the system to accommodate case agnosticism.